### PR TITLE
Fix MUMPS on macOS

### PR DESCRIPTION
**Pull request by Ig-dolci**
# Description

Additional flag to fix macOS issue involving MUMPS.

### CODE REVIEW

**Comment by Ig-dolci** [comment]
Could anyone with macOS help verify if this `firedrake-install` script fixes the parallel issue with MUMPS? 

Please try running `test_poisson_analytic_linear_parallel()` from `tests/regression/test_poisson_strong_bcs.py`. My machine works correctly with `mpiexec -n 2 tests/regression/test_poisson_strong_bcs.py`, but fails when run with `pytest`.

If the pytest also fails for others, it might indicate a parallel issue with pytest rather than with MUMPS. 

You can try pytest and python executions for `test_dg_advection_cubed_sphere_parallel()` from `tests/regression/test_dg_advection.py`. 

Obs: Without `-Wl,-dead_strip_dylibs` flag, neither python nor pytest executions work.

**Comment by Ig-dolci** [comment]
We also have this Colin test: 

```from firedrake import *
from pyop2.mpi import MPI

size = MPI.COMM_WORLD.size
ensemble = Ensemble(COMM_WORLD, size//2)
rank = ensemble.ensemble_comm.rank

mesh = UnitSquareMesh(20, 20, comm=ensemble.comm)
V = FunctionSpace(mesh, "CG", 1)

n_Js = [3, 3]
Js_offset = [0, 3]

u = TrialFunction(V)
v = TestFunction(V)
a = inner(grad(u), grad(v))*dx

for i in range(n_Js[rank]):
    val = Js_offset[rank]+i+1
    x = Function(V)
    L = x*v*dx
    u0 = Function(V)
    bc = DirichletBC(V, Constant(val), "on_boundary")
    solve(a == L, u0, bcs=[bc])
    print(rank, norm(u0), "NOOORMM")
```
with correct output: 
```
 ~  mpiexec -n 2 python3 mfe.py                                                                                                               
1 3.9999999999999716 NOOORMM
1 5.0 NOOORMM
1 6.000000000000031 NOOORMM
0 0.9999999999999929 NOOORMM
0 1.9999999999999858 NOOORMM
0 3.0000000000000155 NOOORMM
 ~  mpiexec -n 4 python3 mfe.py                                                                                                               
0 0.9999999999999949 NOOORMM
0 1.9999999999999898 NOOORMM
0 2.999999999999998 NOOORMM
0 0.9999999999999949 NOOORMM
0 1.9999999999999898 NOOORMM
0 2.999999999999998 NOOORMM
1 3.9999999999999796 NOOORMM
1 5.0 NOOORMM
1 5.999999999999996 NOOORMM
1 3.9999999999999796 NOOORMM
1 5.0 NOOORMM
1 5.999999999999996 NOOORMM
```

File: scripts/firedrake-install [hunk 1]
```diff
@@ -780,14 +780,14 @@ def get_petsc_options(minimal=False):
             petsc_options.add("--CFLAGS=-Wno-implicit-function-declaration")
             if Version(clt["version"]) >= Version("15"):
                 # CLT >= 15 requires legacy linking behaviour
-                petsc_options.add("--LDFLAGS=-Wl,-ld_classic")
+                petsc_options.add("--LDFLAGS=-Wl,-ld_classic -Wl,-dead_strip_dylibs")
```
Comment (JDBetteridge) [inline scripts/firedrake-install:783]: I think this should be:
```suggestion
                petsc_options.add("--LDFLAGS=-Wl,-ld_classic,-dead_strip_dylibs")
```
and likewise below.

File: scripts/firedrake-install [hunk 2]
```diff
@@ -780,14 +780,14 @@ def get_petsc_options(minimal=False):
...
     elif options.get("with_blas") is not None:
         petsc_options.add("--with-blaslapack-dir={}".format(options["with_blas"]))
         if osname == "Darwin":
             petsc_options.add("--CFLAGS=-I{}/include -Wno-implicit-function-declaration".format(options["with_blas"]))
             if Version(clt["version"]) >= Version("15"):
                 # CLT >= 15 requires legacy linking behaviour
-                petsc_options.add("--LDFLAGS=-Wl,-ld_classic,-rpath,{0}/lib -L{0}/lib".format(options["with_blas"]))
+                petsc_options.add("--LDFLAGS=-Wl,-ld_classic -Wl,-dead_strip_dylibs,-rpath,{0}/lib -L{0}/lib".format(options["with_blas"]))
```
Comment (JDBetteridge) [inline scripts/firedrake-install:790]:
```suggestion
                petsc_options.add("--LDFLAGS=-Wl,-ld_classic,-dead_strip_dylibs,-rpath,{0}/lib -L{0}/lib".format(options["with_blas"]))
```

File: scripts/firedrake-install [hunk 3]
```diff
@@ -797,7 +797,7 @@ def get_petsc_options(minimal=False):
         petsc_options.add("--CFLAGS=-I{}/include -Wno-implicit-function-declaration".format(brew_blas_prefix))
         if Version(clt["version"]) >= Version("15"):
             # CLT >= 15 requires legacy linking behaviour
-            petsc_options.add("--LDFLAGS=-Wl,-ld_classic,-rpath,{0}/lib -L{0}/lib -L{1}".format(brew_blas_prefix, brew_gcc_libdir()))
+            petsc_options.add("--LDFLAGS=-Wl,-ld_classic -Wl,-dead_strip_dylibs,-rpath,{0}/lib -L{0}/lib -L{1}".format(brew_blas_prefix, brew_gcc_libdir()))
```
Comment (JDBetteridge) [inline scripts/firedrake-install:800]:
```suggestion
            petsc_options.add("--LDFLAGS=-Wl,-ld_classic,-dead_strip_dylibs,-rpath,{0}/lib -L{0}/lib -L{1}".format(brew_blas_prefix, brew_gcc_libdir()))
```

**Review by JDBetteridge — CHANGES_REQUESTED** [review]
I also think these options should be passed directly to MUMPS if it is the only package which needs `-Wl,-dead_strip_dylibs`, rather than passing the option globally to PETSc. However, we already have a million random hacks in for MacOS (because things "just work" on a Mac :roll_eyes: ) and I have no interest in dragging this PR out.

Thanks for your continued dedication to providing MacOS fixes!

**Comment by ksagiyam** [comment]
After a fresh install on my M3 Mac, parallel `pytest` also fails for me, but running  `test_poisson_analytic_linear_parallel()` directly with `mpiexec -n 2 python ...` works. Colin's test also passes.

**Comment by Ig-dolci** [comment]
> After a fresh install on my M3 Mac, parallel `pytest` also fails for me, but running `test_poisson_analytic_linear_parallel()` directly with `mpiexec -n 2 python ...` works. Colin's test also passes.

So, I believe we have two options here:

1. Merge this PR and open an issue for pytest fails.
2. Or try to figure out the pytest issue, and fix it before merging this PR.

Any thoughts?

**Comment by ksagiyam** [comment]
`1`, I think, if this PR fixes a major issue.

**Review by ksagiyam — DISMISSED** [review]

File: scripts/firedrake-install [hunk 4]
```diff
@@ -780,14 +780,14 @@ def get_petsc_options(minimal=False):
             petsc_options.add("--CFLAGS=-Wno-implicit-function-declaration")
             if Version(clt["version"]) >= Version("15"):
                 # CLT >= 15 requires legacy linking behaviour
-                petsc_options.add("--LDFLAGS=-Wl,-ld_classic")
+                petsc_options.add("--LDFLAGS=-Wl,-ld_classic,-dead_strip_dylibs")
```
Comment (dham) [inline scripts/firedrake-install:783]: Please add a comment pointing out that this option has been cargo culted.

**Review by dham — DISMISSED** [review]

File: scripts/firedrake-install (hunk 4 above)
Comment (Ig-dolci) [reply scripts/firedrake-install:786]: done.

**Review by JDBetteridge — APPROVED** [review]

**Pull request merged**